### ISSUE TITLE
Feature request: Optionally exclude test classes from scan

### ISSUE THREAD

**Issue by smillies** [lukehutch/fast-classpath-scanner#79]
I am dynamically creating a whitelist of classes from some packages in my Maven project. Of course, in the development environment the project also has the test classes on the classpath. However, the whitelist should not include anything under target/test/...

It would be nice if there were some way to limit scanning to specific _directories_ on the classpath (e. g. to target/main/classes), just as there is a syntax to limit scanning to jar files.

-- Sebastian

**Comment by lukehutch** [comment]
Thanks for this report, this is an interesting request, and I'll have to figure out the best way to implement this since the scan spec is already too complex! A workaround is to use `FastClasspathScanner#overrideClasspath(...)` before calling `#scan()`.

I'll take a look at this ASAP, although my responses may be delayed over the next couple of weeks.

**Comment by lukehutch** [comment]
Hi Sebastian, I'm trying to figure out if this is worth making changes for (e.g. allowing the user to add a directory blacklist regexp before starting the scan). However, it seems that there should be another way to solve this problem for your usecase. 

The test classpath (including target/test/) is only visible to tests. The main downside to having those in the class is that the scanning takes a bit longer (scanning will only take marginally longer if there is a package whitelist specified that does not contain the unit test packages; scanning could take quite a bit longer if there are a large number of large test classes). However, it shouldn't actually break the tests having target/test/ on the classpath, if the scan spec is set up correctly. Right?

(In the main Maven scope, target/main is visible but target/test is not.)

Please explain a bit more what you're trying to avoid here by limiting scanning of certain directories on the classpath. Thanks.

**Comment by smillies** [comment]
Yes, I think you're right. The one complication is when I create whitelists for a productive system not at runtime but during the Maven Build, and target/test is visible. As a matter of principle, you don't want test classes in production code, not even their names. And you cannot exclude unit test packages based on their names, after all, they are the same names as the productive packages.

There are several ways to work around that: overriding the classpath of the scanner, as you mentioned, or calling an Ant task with a modified classpath, and probably others. 

Perhaps easier than introducing directory regexes would be to exclude classes with a name matching "*Test". There is already something that can do that: the `FastClasspathScanner#matchFilenamePattern` family of methods. It would be nice if one could use this mechanism for matching only, instead of having to pass in a `FileMatchProcessor` that actually opens a stream to the file.

**Comment by lukehutch** [comment]
If Maven is specifically the problem, have you tried the `<scope>` element? I know it works for dependencies (you can limit certain dependencies just to the test scope), maybe that solves the problem? Or maybe you can set up two Maven builds, one for test, one for production? It seems like the bug here might be in your Maven setup, although I'm not a Maven expert.

One other thing I thought of was to add regexp matching to the scan spec. I'd prefer to see first if there is some other way to solve this particular issue by fixing the Maven configuration or something, but if there really is no other way, let me know.

**Comment by lukehutch** [comment]
@smillies: Ping :-)  I will be making a release in the near future, and I want to resolve this issue first. Please see my previous comment. Thanks!

**Comment by smillies** [comment]
I couldn't really find a way to fix the Maven config, but it's possible to call an Ant task from Maven that excludes some elements from the classpath. Works for me. Sorry for not having got back to you earlier.

**Comment by lukehutch** [comment]
@smilies: thanks for the reply, and I'm glad you found a workaround, but I'm trying to determine if this is a fix that will be generally useful to others too, without unnecessarily complicating the scanning specification. 

I guess the thing I don't understand is why having the test classes visible on the classpath is actually a problem. While the tests are running, you need the test classes visible on the classpath, and if you want to exclude them, you just whitelist the packages you want to scan. If the tests are not running (i.e. in the non-test scope), Maven doesn't include the test classes in the classpath. What am I missing?

**Comment by smillies** [comment]
Hm, I don't know how to explain that any better as I already tried to do. Let me try again: 

1. I'm creating a whitelist (for deserialization) at build time. 
2. At that time, the test classes are on the classpath.
3. I do not want the test class names on the whitelist (although perhaps they wouldn't do any harm, but they are just noise, and would prevent pre-loading all whitelisted classes at runtime).
4. I cannot exclude the test classes by package, because they sit in the same packages as the productive classes. 
5. I cannot restrict the scan to jars, because packaging has not yet happened. (Indeed, I want the generated whitelist to be packaged together with the productive code later). 

So I wanted to have a means to remove the test class directory from the classpath that is scanned at build time.

**Comment by johnou** [comment]
Sounds like your build pipeline is tightly coupled or back to front? You might consider creating the whitelist after you have created the jars but before building the deployment artifact / package. If your deploy / packaging module which could create the whitelist references your other modules the test classes won't be on the classpath.

**Comment by smillies** [comment]
I hadn't thought of that. Anyway, I'm sure there are several ways to do it that don't involve tampering with FastClasspathScanner.

**Comment by lukehutch** [comment]
You can already blacklist specific classes using `"-com.package.MyUnitTest"` in the scan spec. I'm just trying to figure out if adding something like `"-com.package.*Test"` would be worth the complexity. (Adding wildcards to the scanning process will significantly complicate the recursive directory scanning logic.)

You want to exclude all classes whose name ends with `Test`. Can't you just filter the list of classnames that are returned by FastClasspathScanner to exclude classes matching this pattern? For example:

```java
List<String> widgetClasses =
    new FastClasspathScanner("com.xyz.widget").scan()
        .getClassNameToClassInfo().values().stream()
            .filter(ci -> !ci.getClassName().endsWith("Test"))
            .map(ClassInfo::getClassName)
            .sorted()
            .collect(Collectors.toList());
```

**Comment by smillies** [comment]
Yes, I could do that. There are also some test classes that are inappropriately named, and of course test utilities that do not fall under this naming convention, but hard-coding a small explicit blacklist wouldn't hurt. (One would need to remember to edit this explicit list whenever there is a new test helper, or put them all into certain packages.)

**Comment by lukehutch** [comment]
Is there another way you can detect test classes, e.g. using class annotations? You can also filter for that.

Given the above discussion, I really think some form of post-scan filtering is the right approach for solving this, but I'm still open to be convinced otherwise.

**Comment by lukehutch** [comment]
Closing, since post-scan filtering seems to be the right way to solve this in the general case.
